cmake/traceparser: warn user for too old cmake

gerion0 opened the pull request:
CMake <3.17 is error prone in some cases. Warn the user about that.

I'm not entirely happy with this commit, since it results in a lot of warnings, because `_guess_files` is often called multiple times.
However, `_guess_files` is the place, where the errors can occur and Meson needs to guess and the warning will appear less and less in the future.

See also: #7601

mensinda commented:
I don't think that a warning is reasonable here since there are a lot of projects that work (more or less) fine with CMake <3.17. A short note (as in `mlog.log`) when selecting the CMake executable in `executor.py` which references the (also updated docs with more details) would be more reasonable.

You can print a warning/log only once by passing `once=True` as a kwarg to the `mlog` functions.

gerion0 commented:
> I don't think that a warning is reasonable here since there are a lot of projects that work (more or less) fine with CMake <3.17. A short note (as in `mlog.log`) when selecting the CMake executable in `executor.py` which references the (also updated docs with more details) would be more reasonable.

I have changed the log level to log instead of warning and have extended the documentation. However, I think executor is the wrong place for the message. First there already is a check for 3.17 in traceparser that needs to be dublicated, when warn in executor and second is the point where the guessing happens the `_guess_files` function in traceparser. So as a Meson developer (or anyone who looks through the code) it is much more understandable, if the warning for ambiguous parsing is emitted in exactly this part where the guessing is done.

> You can print a warning/log only once by passing `once=True` as a kwarg to the `mlog` functions.

Actually not. All other log levels of `mlog` have the `once` kwarg, but `log` has another function `log_once`. Anyway, I have used it.

mensinda commented:
> So as a Meson developer (or anyone who looks through the code) it is much more understandable, if the warning for ambiguous parsing is emitted in exactly this part where the guessing is done.

The first problem with this is that `_guess_files` is not the only place where different code paths based on CMake versions are executed. This is because the CMake trace format is not the only CMake feature that has changed. There is, for instance, `--trace-redirect` which was introduced in CMake 3.16 and the CMake file API which was introduced in CMake 3.14 (see the [`CMakeInterpreter`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/cmake/interpreter.py#L848)).

The second problem is that this message will also be printed for CMake dependencies. There `_guess_files` is far less fragile since the file paths are far more predictable and have file extensions in 99.9% of all cases. So, printing this message there would not be helpful for the end-user.

Also, as an end-user, I would expect meson to warn/notify me about CMake version requirements close to where CMake was detected and not after calling CMake and somewhere in the middle of the logs for processing a CMake dependency/subproject.

Just hardcoding CMake 3.17 in `CMakeExecutor` might indeed not have been the best call, however, the `CMakeExecutor` could take a new (optional) `recommended_version` argument which would be more flexible. Alternatively, the version check can also be put in the `CMakeInterpreter` with the other version checks.

eli-schwartz commented:
This is obsolete since #9906 because we now warn you for cmake <3.17 everywhere.
